Fix undefined match results in "undefined" string

sapkra opened the pull request:
This should fix a bug that when you have a matcher which should show an empty string with some styles it results in showing an `undefined` string.

Before:
```
{{{matcher0}}}undefined{{{/matcher0}}}
```
After:
```
{{{matcher0}}}{{{/matcher0}}}
```

### Example:
```
>Hello
>
>World
```

>Hello
>undefined
>World


>Hello
>
>World

milesj commented:
@sapkra The `match` property is required to be returned from the matcher. What are you currently returning?

sapkra commented:
I'm just returning index 1 of the matches array to remove the `>` like you did it in the examples.

btw. I have to generate a key otherwise Interweave will produce errors but it was also the case in v11.

```
export class MarkdownQuoteMatcher extends Matcher<any> {
  public replaceWith(children: ChildrenNode): Node {
    return <blockquote key={generate()}>{children}</blockquote>;
  }

  public asTag() {
    return "blockquote";
  }

  public match(value: string) {
    return this.doMatch(value, /^(?:>)(.*)?/u, (matches) => ({ match: matches[1] }));
  }
}
```

milesj commented:
@sapkra The parser sets `key` for you: https://github.com/milesj/interweave/blob/master/packages/core/src/Parser.ts#L185 You need to spread the props in that are passed as the 2nd argument to `replaceWith`.

Are you sure `matches[1]` isn't undefined?

sapkra commented:
You are right. The index 1 of the match object is undefined. But undefined should never be rendered as a string so I still think that this is a bug of v12 because such errors should be catched by the library itself and not in all Matchers and I didn't had this problem in v11.

The matches array:
```javascript
0: ">"
1: undefined
groups: undefined
index: 0
input: ">"
length: 2
```

Related to the problem with the key. I didn't realized that these are already generated - so thanks for noticing it. I previously tried to do it like in your example by completely adding the props object of `replaceWith` to the rendered element but then I had the problem that all props I've added to the `BaseInterweave` component are also added to the element in my matcher (it should be like that I guess). So I will spread it like you said and just use the key out of the props.

milesj commented:
@sapkra The matcher should be returning the correct response. I don't agree that undefined is a correct response, as you're then creating a start and close token pair with no inner content. If a match is unsuccessful, you should maybe just return `null`?

I suppose I need a bigger example that the one in the original post.

sapkra commented:
Ok I understand but at this point we have just different opinions. My point is that a library should be smart enough to catch those errors without letting the user think about all edge cases and fix possible errors by themselve.
But you're the library author and it's of cause your decision how you want to handle it. I will close the PR and just add a fallback to my matcher.
